tests/posix/pthread_key/src/pthread_key.c

```diff
@@ -61,45 +92,111 @@ static void make_key(void)
 	zassert_false(ret, "insufficient memory to create key");
 }
 
+static void make_keys(void)
+{
+	int i, ret = 0;
+
+	for (i = 0; i < N_KEY; i++) {
+		ret = pthread_key_create(&keys[i], NULL);
+		zassert_false(ret, "insufficient memory to create keys");
+	}
+}
+
 /**
  * @brief Test to demonstrate pthread_key APIs usage
  *
- * @details The test spawns a thread which uses pthread_once() to
+ * @details The tests spawn a thread which uses pthread_once() to
  * create a key via pthread_key_create() API. It then sets the
- * thread-specific value to the key using pthread_set_specific() and
- * gets it back using pthread_get_specific and asserts that they
+ * thread-specific value to the key using pthread_setspecific() and
+ * gets it back using pthread_getspecific and asserts that they
  * are equal. It then deletes the key using pthread_key_delete().
+ * Both the sub-tests do the above, but one with multiple threads
+ * using the same key and the other with a single thread using
+ * multiple keys.
  */
 
-void test_pthread_key(void)
+void test_multiple_threads_single_key(void)
 {
-	int ret = -1;
+	int i, ret = -1;
 
-	pthread_attr_t attr;
+	pthread_attr_t attr[N_THR];
 	struct sched_param schedparam;
-	pthread_t newthread;
+	pthread_t newthread[N_THR];
+	void *retval;
 
 	ret = pthread_once(&key_once, make_key);
 
 	/* TESTPOINT: Check if key is created */
 	zassert_false(ret, "attempt to create key failed");
 
+	printk("\nDifferent threads set different values to same key:\n");
+
+	/* Creating threads with lowest application priority */
+	for (i = 0; i < N_THR; i++) {
+		pthread_attr_init(&attr[i]);
+		schedparam.priority = 2;
+		pthread_attr_setschedparam(&attr[i], &schedparam);
+		pthread_attr_setstack(&attr[i], &stacks[i][0], STACKSZ);
+
+		ret = pthread_create(&newthread[i], &attr[i], thread_top,
+				(void *)i);
+
+		/* TESTPOINT: Check if threads are created successfully */
+		zassert_false(ret, "attempt to create threads failed");
+	}
+
+	for (i = 0; i < N_THR; i++) {
+		printk("thread %d: ", i);
+		pthread_join(newthread[i], &retval);
+	}
+
+	ret = pthread_key_delete(key);
+
+	/* TESTPOINT: Check if key is deleted */
+	zassert_false(ret, "attempt to delete key failed");
+	printk("\n");
+}
+
+void test_single_thread_multiple_keys(void)
+{
+	int i, ret = -1;
+
+	pthread_attr_t attr;
+	struct sched_param schedparam;
+	pthread_t newthread;
+
+	ret = pthread_once(&keys_once, make_keys);
+
+	/* TESTPOINT: Check if keys are created successfully */
+	zassert_false(ret, "attempt to create keys failed");
+
+	printk("\nSingle thread associates its value with different keys:\n");
 	pthread_attr_init(&attr);
```

Comment:
Check return value of `attr_init` before calling `setschedparam` or `setstack`, fixes #9297 (unchecked return value)